[Donovan Neo] iP

xenthm opened the pull request.

xenthm committed:
Rename to Pat

Only done for main source file for now.

xenthm committed:
Decompose Pat's messages and add colour

Move Pat's greeting and goodbye into their own methods for modularity.

Add colour to Pat's messages to differentiate from the user.

xenthm committed:
Echo user messages

Add ANSI escape character sequences to manipulate the output of the app
for a cleaner user experience. This only works in the native OS terminal.

xenthm committed:
Add add task to todo and list tasks methods

xenthm committed:
Add mark/unmark task method and refactor task implementation

Take advantage of OOP and abstract away Task into its own class.

xenthm committed:
Reformat Pat's outputs

Get Pat to respond more verbosely to improve user expereince.

xenthm committed:
Add inherited classes for polymorphism

Add Todo, Event, and Deadline classes for more fine-tuned behaviour.

xenthm committed:
Add cases in Main for subclasses of Task

Also give usage hints when user types a wrong command.

xenthm committed:
Add text UI testing with redirection

xenthm committed:
Update UI test expected output

xenthm committed:
Refactor to improve code quality

- Define "magic" ANSI escape character sequences as constants.
- Reduce levels of abstraction in the main chat loop by extracting
methods to improve code readability and keep to SLAP guidelines.

xenthm committed:
Update documentation to Pat

yentheng0110 commented on src/main/java/Pat.java:
```diff
@@ -0,0 +1,172 @@
+    public static final String ANSI_RESET = "\033[0m";
+    public static final String ANSI_ERASE_LINE = "\033[2K";
+    public static final String ANSI_MOVE_TO_START_OF_PREVIOUS_LINE = "\033[F";
+    public static final String INDENT = "    ";
```
I like how you extracted these constants to improve code quality

yentheng0110 commented on src/main/java/Pat.java:
```diff
@@ -0,0 +1,172 @@
+        say("Added this task: ");
+        say(INDENT + task.getTask());
+        int size = todoList.size();
+        say("Now you have " + size + (size == 1 ? " task" : " tasks") + " in the list");
```
I like how you notice the difference in singular vs plural forms of task. Perhaps can extract this (size == 1 ? " task" : " tasks") as a String variable

yentheng0110 commented on src/main/java/Pat.java:
```diff
@@ -0,0 +1,172 @@
+        } while (!saidBye);
+    }
+
+    public static void main(String[] args) {
```
Good use of SLAP coding quality

yentheng0110 commented on src/main/java/Pat.java:
```diff
@@ -0,0 +1,172 @@
+        }
+    }
+
+    private static void handleDeadline(String arg) {
+        try {
+            String[] splitLineBy = arg.split(" /by ", 2);
+            String deadline = splitLineBy[0].trim();
+            String by = splitLineBy[1].trim();
+            if (deadline.isEmpty()) {
+                throw new IllegalArgumentException("Empty deadline");
+            } else if (by.isEmpty()) {
+                throw new IllegalArgumentException("Empty by");
+            }
+            addToTodo(new Deadline(deadline, by));
+        } catch (IllegalArgumentException | IndexOutOfBoundsException e) {
+            say("Please give me a valid deadline to add!");
+            say("deadline [task] /by [deadline]");
+        }
+    }
+
+    private static void handleEvent(String arg) {
+        try {
+            String[] splitLineFrom = arg.split(" /from ", 2);
+            String event = splitLineFrom[0].trim();
+            String[] splitLineTo = splitLineFrom[1].split(" /to ", 2);
+            String from = splitLineTo[0].trim();
+            String to = splitLineTo[1].trim();
+            if (event.isEmpty()) {
+                throw new IllegalArgumentException("Empty event");
+            } else if (from.isEmpty()) {
+                throw new IllegalArgumentException("Empty from");
+            } else if (to.isEmpty()) {
+                throw new IllegalArgumentException("Empty to");
+            }
+            addToTodo(new Event(event, from, to));
+        } catch (IllegalArgumentException | IndexOutOfBoundsException e) {
+            say("Please give me a valid event to add!");
+            say("event [task] /from [start] /to [end]");
+        }
+    }
```
I like how you extracted these methods to improve your code quality. Good use of try and catch exceptions to handle invalid commands from the user.

yentheng0110 reviewed (commented):
LGTM. Overall your code is clean and easy to read.

Paulifyer commented on src/main/java/Pat.java:
```diff
@@ -0,0 +1,172 @@
+
+    private static void handleEvent(String arg) {
+        try {
+            String[] splitLineFrom = arg.split(" /from ", 2);
+            String event = splitLineFrom[0].trim();
+            String[] splitLineTo = splitLineFrom[1].split(" /to ", 2);
```
Magic Numbers are generally not recommended for the sake of clarity, maybe use constants instead?

Paulifyer commented on src/main/java/Deadline.java:
```diff
@@ -0,0 +1,13 @@
+        this.by = by;
+    }
+
+    @Override
+    public String getTask() {
+        return "[D]" + super.getTask() + " (by: " + by + ")";
+    }
```
I like your consistent usage of  the Override tag

Paulifyer commented on src/main/java/Pat.java:
```diff
@@ -0,0 +1,172 @@
+        }
+    }
+
+    private static void unmarkTask(String arg) {
+        try {
+            Task task = todoList.get(Integer.parseInt(arg) - 1);
+            task.markNotDone();
+            say("OK, I've marked this task as not done yet:");
+            say(task.getTask());
+        } catch (NumberFormatException | IndexOutOfBoundsException e) {
+            say("Please tell me a valid task to unmark!");
+            say("unmark [task number from list]");
+        }
+    }
```
Good use of SLAP to abstract the marking of tasks

Paulifyer commented on src/main/java/Todo.java:
```diff
@@ -0,0 +1,10 @@
+public class Todo extends Task {
+    public Todo(String description) {
+        super(description);
+    }
```
Good following of Naming conventions across the board

xenthm committed:
Add CMD file for building and running Pat

xenthm committed:
Add ANSI escape code constant and move trim() calls

- Add ANSI escape code to move to next line (this was missed out when defining these constants in a previous commit)
- Move trim() calls on lines read with Scanner to minimise its usage

xenthm committed:
Refactor IndexOutOfBoundsException handling

Move it outside individual methods out to runChat()

xenthm committed:
Expand say functionality and revert back handling empty arguments

- Modify say() ot mirror print()
- Add sayln() to mirror println()
- Move handling of empty arguments back to respective methods

xenthm committed:
Define custom exceptions

- InvalidCommandException: for when the user command is badly formed
- EmptyDateException: for when date arguments (such as /by, /from, and /to) are not stated

xenthm committed:
Implement custom exceptions

xenthm committed:
Merge branch 'branch-Level-5'

xenthm committed:
Split classes into packages

- Created pat.exceptions & pat.types to house all classes except Pat.java
- Update Pat.cmd to only run Pat, not build (build in IntelliJ instead)

xenthm committed:
Merge branch 'branch-A-Packages'

xenthm committed:
Fix small bug in response to invalid task number

xenthm committed:
Add delete task method and fix minor bugs

Fix minor bugs in unmark method.

xenthm committed:
Fix minor bug in unmark method

xenthm committed:
Add getters and setters for all task types

xenthm committed:
Add write to file method

- Create txt file with format [T/D/E] followed by other fields in each class, separated by a "|"
- Add data directory to .gitignore
- Update runtest.bat

xenthm committed:
Add constructor for Task and subclasses that takes isDone parameter

xenthm committed:
Add list saving functionality

Save list into .txt file.

xenthm committed:
Merge branch 'branch-Level-6'

xenthm committed:
Merge branch 'branch-Level-7'

xenthm committed:
Add MANIFEST.MF for jar building

xenthm committed:
Change behaviour when no task number supplied to mark/unmark/delete

xenthm committed:
Extract Ui and TaskList classes from Pat

- Ui: deals with interactions with the user
- TaskList: extends ArrayList; contains operations to add, mark, unmark, delete, and print tasks

xenthm committed:
Extract Storage class from Pat

Deals with loading tasks from the file and saving tasks in the file.

xenthm committed:
Extract readCommand method from Pat to Ui class

xenthm committed:
Add constructors to Deadline and Event

These constructors take in String arrays.

xenthm committed:
Extract Parser class from Pat

xenthm committed:
Extract methods from Pat and TaskList classes to Parser

- Make Parser handle all commands for now
- Rename tasklist to taskList

xenthm committed:
Add function to find task

xenthm committed:
Rearrange constructors and methods in classes

Deadline and Event classes.

xenthm committed:
Merge pull request #1 from xenthm/branch-Level-9

Level 9 Increment

xenthm committed:
Make Pat fields private final

xenthm committed:
Add JavaDoc for Pat class

xenthm committed:
Add JavaDoc for TaskList class

xenthm committed:
Remove unused setters in Task class

xenthm committed:
Merge branch 'master' into branch-A-JavaDoc

xenthm committed:
Add JavaDoc for Task class and make constructors protected

xenthm committed:
Remove unused setters in Deadline and Event classes

xenthm committed:
Merge branch 'master' into branch-A-JavaDoc

xenthm committed:
Add JavaDoc for Todo, Deadline, and Event classes

xenthm committed:
Add JavaDoc for Ui class

xenthm committed:
Change RuntimeException to SecurityException in Storage class

xenthm committed:
Merge branch 'master' into branch-A-JavaDoc

xenthm committed:
Add JavaDoc for Storage class

xenthm committed:
Add JavaDoc for Parser class

xenthm committed:
Merge pull request #2 from xenthm/branch-A-JavaDoc

A-JavaDoc increment

xenthm committed:
Fix bug when user sends an empty line

Pat will correctly not do anything. Previously, Pat would tell the user
to send a valid command.

xenthm committed:
Allow Pat to understand mark/unmark/delete commands with extra spaces

xenthm committed:
Update Storage class

- Update access modifiers
- Fix minor errors in JavaDoc and naming

xenthm committed:
Fix bug in the "bye" command

Used to respond as if an invalid command was given. Now, it correctly
doesn't.

xenthm committed:
Add User Guide by updating README.md

Uploaded screenshot of Pat in action as well.

xenthm committed:
Merge pull request #5 from xenthm/branch-A-UserGuide

A-UserGuide increment

xenthm committed:
Remove rendered alert portion from README.md

There is a bug that prevents it from rendering correctly.

xenthm committed:
Update MANIFEST.MF for release v0.2

xenthm committed:
Save task list after every command that modifies it

- Move the call to write the task list to the data file into
  Parser::parseCommand
- Update JavaDoc for Parser::parseCommand

xenthm committed:
Update README.md to reflect save list functionality

xenthm committed:
Update MANIFEST.MF for release v0.3

xenthm committed:
Update README.md

xenthm committed:
Change file structure so that "pat" is a package

Initially, "pat" was set as the source folder of the project, and each
of the top-level classes (Pat, Parser, etc.) were not part of a package.

This commit sets "src" as the source folder, and made all the top-level
classes part of the package "pat".

Pat.cmd was also updated to ensure Pat can be run with a short command.

xenthm committed:
Merge branch 'master' of https://github.com/xenthm/ip

xenthm committed:
Update MANIFEST.MF to reflect new package structure

xenthm committed:
Add Checkstyle from tP

As per issue #22 on course forum
